core/trace/src/main/scala/org/typelevel/otel4s/trace/TraceScope.scala

```diff
@@ -0,0 +1,339 @@
+    * │   Context                                    │       │   Context                                    │
+    * │                                              │       │                                              │
+    * │  ┌─────────────────┐    ┌─────────────────┐  │       │  ┌─────────────────┐    ┌─────────────────┐  │
+    * │  │                 │    │                 │  │       │  │                 │    │                 │  │
+    * │  │   SpanContext   │    │     Baggage     │  │  ---> │  │   SpanContext   │    │     Baggage     │  │
+    * │  │ span_id = "abc" │    │                 │  │       │  │     invalid     │    │                 │  │
+    * │  └─────────────────┘    └─────────────────┘  │       │  └─────────────────┘    └─────────────────┘  │
+    * │                                              │       │                                              │
+    * └──────────────────────────────────────────────┘       └──────────────────────────────────────────────┘
+    * }}}
+    *
+    * ===The [[SpanContext]] is invalid===
+    *
+    * The current propagation strategy is already no-op, so we keep the context
+    * as is. The baggage data remains.
+    *
+    * {{{
+    * ┌──────────────────────────────────────────────┐       ┌──────────────────────────────────────────────┐
+    * │   Context                                    │       │   Context                                    │
+    * │                                              │       │                                              │
+    * │  ┌─────────────────┐    ┌─────────────────┐  │       │  ┌─────────────────┐    ┌─────────────────┐  │
+    * │  │                 │    │                 │  │       │  │                 │    │                 │  │
+    * │  │   SpanContext   │    │     Baggage     │  │  ---> │  │   SpanContext   │    │     Baggage     │  │
+    * │  │     invalid     │    │                 │  │       │  │     invalid     │    │                 │  │
+    * │  └─────────────────┘    └─────────────────┘  │       │  └─────────────────┘    └─────────────────┘  │
+    * │                                              │       │                                              │
+    * └──────────────────────────────────────────────┘       └──────────────────────────────────────────────┘
+    * }}}
+    */
+  def noopScope: F ~> F
+
+  /** Creates a scope that uses the given `context`.
+    *
+    * A shortcut for
+    * {{{
+    * Local[F, Context].scope(context)
+    * }}}
+    *
+    * @param context
+    *   the context to use
+    */
+  def withContext(context: Ctx): F ~> F
+
+  /** Runs the `f` with the current [[Ctx]].
+    *
+    * A shortcut for
+    * {{{
+    * Local[F, Context].reader(f)
+    * }}}
+    */
+  def contextReader[A](f: Ctx => A): F[A]
+}
+
+private[otel4s] object TraceScope {
+
+  def fromLocal[F[_], Ctx, K[X] <: Key[X]](
+      getSpanContext: Ctx => Option[SpanContext],
```

Comment:
I've tried using the `K[SpanContext]` here, but it does not work for oteljava implementation because we cannot translate the OpenTelemetry key into our.